server/conn.go

```diff
@@ -643,27 +646,63 @@ func (cc *clientConn) readOptionalSSLRequestAndHandshakeResponse(ctx context.Con
 		return err
 	}
 
-	// switching from other methods should work, but not tested
-	if resp.AuthPlugin == mysql.AuthCachingSha2Password {
-		resp.Auth, err = cc.authSwitchRequest(ctx)
-		if err != nil {
-			logutil.Logger(ctx).Warn("attempt to send auth switch request packet failed", zap.Error(err))
-			return err
-		}
-	}
 	cc.capability = resp.Capability & cc.server.capability
 	cc.user = resp.User
 	cc.dbname = resp.DBName
 	cc.collation = resp.Collation
 	cc.attrs = resp.Attrs
 
+	newAuth, err := cc.checkAuthPlugin(ctx, &resp.AuthPlugin)
+	if err != nil {
+		logutil.Logger(ctx).Warn("failed to check the user authplugin", zap.Error(err))
+	}
+	if len(newAuth) > 0 {
+		resp.Auth = newAuth
+	}
+
+	switch resp.AuthPlugin {
+	case mysql.AuthCachingSha2Password:
+		resp.Auth, err = cc.authSha(ctx)
+	case mysql.AuthNativePassword:
+	default:
+		return errors.New("Unknown auth plugin")
+	}
```

Comment:
You could potentially use an `authSwitchRequest` (from deleted code above) if the plugin is unknown. The downside is that this is hard to test.

Reply:
If the client presents an authentication plugin that doesn't match the plugin used by the server `checkAuthPlugin` will do the `authSwitchRequest` so I don't think this code will be reached during normal operation. It may be reached if someone bypasses the `GRANT`/`REVOKE` etc and changes `mysql.user` directly.